digdag-core/src/main/java/io/digdag/core/database/BasicDatabaseStoreManager.java

```diff
@@ -380,62 +455,78 @@ private void validateTransactionAndCommit(Handle handle)
 
     public <T> T autoCommit(AutoCommitAction<T, D> action)
     {
-        try {
-            return transactionRetryExecutor.runInterruptible(() -> {
-                try (Handle handle = dbi.open()) {
-                    return action.call(handle, handle.attach(daoIface));
-                }
-            });
-        }
-        catch (RetryGiveupException ex) {
-            Throwable innerException = ex.getCause();
-            throw Throwables.propagate(innerException);
+        Handle nested = currentTransaction.get();
+        if (nested != null) {
+            return action.call(nested, nested.attach(daoIface));
         }
-        catch (InterruptedException ex) {
-            throw Throwables.propagate(ex);
+        else {
+            try {
+                return transactionRetryExecutor.runInterruptible(() -> {
+                    try (Handle handle = dbi.open()) {
```

Comment:
I see.

> Moreover, we shouldn't write such code...do you think BasicDatabaseStoreManager should have a assertion to make nested transaction() in autoCommit failed?

Ideally, it's best to prevent those codes at compile time...